File: app/src/main/java/com/amazon/tv/leanbacklauncher/settings/LegacyAppRowPreferenceFragment.java
package com.amazon.tv.leanbacklauncher.settings;

import android.app.Activity;
import android.content.Intent;
import android.os.Bundle;
import android.support.annotation.NonNull;
import android.support.v17.leanback.app.GuidedStepSupportFragment;
import android.support.v17.leanback.widget.GuidanceStylist;
import android.support.v17.leanback.widget.GuidedAction;
import android.support.v4.content.res.ResourcesCompat;
import android.text.InputType;
import android.util.Log;
import android.widget.Toast;

import com.amazon.tv.firetv.leanbacklauncher.apps.AppCategory;
import com.amazon.tv.firetv.leanbacklauncher.apps.RowPreferences;
import com.amazon.tv.firetv.leanbacklauncher.util.FireTVUtils;
import com.amazon.tv.firetv.leanbacklauncher.util.SharedPreferencesUtil;
import com.amazon.tv.leanbacklauncher.MainActivity;
import com.amazon.tv.leanbacklauncher.R;

import java.lang.Integer;
import java.util.ArrayList;
import java.util.List;
import java.util.Set;


/**
 * Created by rockon999 on 3/25/18.
 */

public class LegacyAppRowPreferenceFragment extends GuidedStepSupportFragment {

    private static final int ACTION_ID_APPS = 50;
    private static final int ACTION_ID_APPS_MIN = 51;
    private static final int ACTION_ID_APPS_MAX = 52;
    private static final int ACTION_ID_RECOMENDATIONS = 100;
    private static final int ACTION_ID_INPUTS = 200;
    private static int startup = 0;

    public void onCreate(Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);
    }

    @NonNull
    public GuidanceStylist.Guidance onCreateGuidance(Bundle savedInstanceState) {
        return new GuidanceStylist.Guidance(getString(R.string.select_app_customize_rows_title), null, getString(R.string.home_screen_order_content_title), ResourcesCompat.getDrawable(getResources(), R.drawable.ic_settings_home, null));
    }

    public void onResume() {
    	super.onResume();
        this.updateActions();
        startup++;
    }

    public void onPause() {
    	super.onPause();
        startup = 0;
    }

    private int musicIndex, gameIndex, videoIndex, favIndex;

    public void onGuidedActionClicked(GuidedAction action) {

        long id = action.getId();
        long modId = (id - 1) / 2;
        int subId = (int) ((id - 1) % 2);
        int val = 1;
        boolean enabled = false;

        Activity activity = getActivity();
        Set<AppCategory> categories = RowPreferences.getEnabledCategories(activity);

        // Log.w("+++ action", action.toString());
        // Log.w("+++ action.id", ""+id);
        // Log.w("+++ modId", " "+modId);
        // Log.w("+++ subId", " "+subId);

        if (modId == favIndex) {
            switch (subId) {
                case 0:
                    enabled = RowPreferences.areFavoritesEnabled(activity);
                    // Log.w("+++ favorites.enabled", ""+enabled);
                    RowPreferences.setFavoritesEnabled(activity, !enabled);
                    break;
                case 1:
                    try {
                        val = Integer.parseInt(action.getDescription().toString());
                    } catch (NumberFormatException nfe) {
                        val = 1;
                    }
                    // Log.w("+++ set fav.max ", ""+val);
                	RowPreferences.setFavoriteRowMax(activity, val);
                	break;
                case 2:
                    try {
                        val = Integer.parseInt(action.getDescription().toString());
                    } catch (NumberFormatException nfe) {
                        val = 1;
                    }
                    // Log.w("+++ set fav.min ", ""+val);
                	RowPreferences.setFavoriteRowMin(activity, val);
            }
        } else if (modId == musicIndex) {
            switch (subId) {
                case 0:
                    enabled = categories.contains(AppCategory.MUSIC);
                    // Log.w("+++ music.enabled", ""+enabled);
                    RowPreferences.setMusicEnabled(activity, !enabled);
                    break;
                case 1:
                    try {
                        val = Integer.parseInt(action.getDescription().toString());
                    } catch (NumberFormatException nfe) {
                        val = 1;
                    }
                    // Log.w("+++ set mucic.max ", ""+val);
                	RowPreferences.setRowMax(AppCategory.MUSIC, activity, val);
                	break;
                case 2:
                    try {
                        val = Integer.parseInt(action.getDescription().toString());
                    } catch (NumberFormatException nfe) {
                        val = 1;
                    }
                    // Log.w("+++ set music.min ", ""+val);
                	RowPreferences.setRowMin(AppCategory.MUSIC, activity, val);
            }
        } else if (modId == videoIndex) {
            switch (subId) {
                case 0:
                    enabled = categories.contains(AppCategory.VIDEO);
                    // Log.w("+++ videos.enabled", ""+enabled);
                    RowPreferences.setVideosEnabled(activity, !enabled);
                    break;
                case 1:
                    try {
                        val = Integer.parseInt(action.getDescription().toString());
                    } catch (NumberFormatException nfe) {
                        val = 1;
                    }
                    // Log.w("+++ set video.max ", ""+val);
                	RowPreferences.setRowMax(AppCategory.VIDEO, activity, val);
                	break;
                case 2:
                    try {
                        val = Integer.parseInt(action.getDescription().toString());
                    } catch (NumberFormatException nfe) {
                        val = 1;
                    }
                    // Log.w("+++ set video.min ", ""+val);
                	RowPreferences.setRowMin(AppCategory.VIDEO, activity, val);
            }
        } else if (modId == gameIndex) {
            switch (subId) {
                case 0:
                    enabled = categories.contains(AppCategory.GAME);
                    // Log.w("+++ games.enabled", ""+enabled);
                    RowPreferences.setGamesEnabled(activity, !enabled);
                    break;
                case 1:
                    try {
                        val = Integer.parseInt(action.getDescription().toString());
                    } catch (NumberFormatException nfe) {
                        val = 1;
                    }
                    // Log.w("+++ set game.max ", ""+val);
                	RowPreferences.setRowMax(AppCategory.GAME, activity, val);
                	break;
                case 2:
                    try {
                        val = Integer.parseInt(action.getDescription().toString());
                    } catch (NumberFormatException nfe) {
                        val = 1;
                    }
                    // Log.w("+++ set game.min ", ""+val);
                	RowPreferences.setRowMin(AppCategory.GAME, activity, val);
            }
        } else if (id == ACTION_ID_APPS_MAX) {
                    try {
                        val = Integer.parseInt(action.getDescription().toString());
                    } catch (NumberFormatException nfe) {
                        val = 1;
                    }
                    // Log.w("+++ set all.max ", ""+val);
                	RowPreferences.setAllAppsMax(activity, val);
        } else if (id == ACTION_ID_APPS_MIN) {
                    try {
                        val = Integer.parseInt(action.getDescription().toString());
                    } catch (NumberFormatException nfe) {
                        val = 1;
                    }
                    // Log.w("+++ set all.min ", ""+val);
                	RowPreferences.setAllAppsMin(activity, val);
        } else if (id == ACTION_ID_RECOMENDATIONS) { // RECOMENDATIONS
            enabled = RowPreferences.areRecommendationsEnabled(activity);
            // Log.w("+++ recommendations.enabled", ""+enabled);
            RowPreferences.setRecommendationsEnabled(activity, !enabled);
            if (!enabled && FireTVUtils.isLocalNotificationsEnabled(activity)) {
            	Toast.makeText(activity, activity.getString(R.string.recs_warning_sale), Toast.LENGTH_LONG).show();
            }
        } else if (id == ACTION_ID_INPUTS) { // INPUTS
            enabled = RowPreferences.areInputsEnabled(activity);
            // Log.w("+++ inputs.enabled", ""+enabled);
            RowPreferences.setInputsEnabled(activity, !enabled);
        }
        updateActions();
    }

    private void updateActions() {

        // Log.w("+++ updateActions()", "+++");
        ArrayList<GuidedAction> actions = new ArrayList<>();
        Activity activity = getActivity();
        Set<AppCategory> categories = RowPreferences.getEnabledCategories(activity);
        String statelabel;
        int i = 0;

        // RECOMENDATIONS
        boolean state = RowPreferences.areRecommendationsEnabled(activity);
        statelabel = (state) ? getString(R.string.v7_preference_on) : getString(R.string.v7_preference_off);
        actions.add(new GuidedAction.Builder(activity).id(ACTION_ID_RECOMENDATIONS).title(R.string.recs_row_title).description(statelabel).build());
        // actions.add(new GuidedAction.Builder(activity).id(ACTION_ID_RECOMENDATIONS).checkSetId(ACTION_ID_RECOMENDATIONS).checked(state).title(statelabel).description("").build());

        // FAV
        state = RowPreferences.areFavoritesEnabled(activity);
        statelabel = (state) ? getString(R.string.v7_preference_on) : getString(R.string.v7_preference_off);
        actions.add(new GuidedAction.Builder(activity).id(++i).title(R.string.favorites_row_title).description(statelabel).build());
        // actions.add(new GuidedAction.Builder(activity).id(++i).checkSetId(i).checked(state).title(statelabel).description("").build());

        int[] constraints = RowPreferences.getFavoriteRowConstraints(activity);

        actions.add(new GuidedAction.Builder(activity).id(++i).title(R.string.max_rows_title).description(Integer.toString(constraints[1])).descriptionEditable(true).descriptionEditInputType(InputType.TYPE_CLASS_NUMBER | InputType.TYPE_NUMBER_FLAG_SIGNED).build());
        // actions.add(new GuidedAction.Builder(activity).id(++i).title(R.string.min_rows_title).description(Integer.toString(constraints[0])).descriptionEditable(true).descriptionEditInputType(InputType.TYPE_CLASS_NUMBER | InputType.TYPE_NUMBER_FLAG_SIGNED).build());

        favIndex = ((i - 1) / 2);

        // MUSIC
        state = categories.contains(AppCategory.MUSIC);
        statelabel = (state) ? getString(R.string.v7_preference_on) : getString(R.string.v7_preference_off);
        actions.add(new GuidedAction.Builder(activity).id(++i).title(R.string.music_row_title).description(statelabel).build());
        // actions.add(new GuidedAction.Builder(activity).id(++i).checkSetId(i).checked(state).title(statelabel).description("").build());

        constraints = RowPreferences.getRowConstraints(AppCategory.MUSIC, activity);

        actions.add(new GuidedAction.Builder(activity).id(++i).title(R.string.max_rows_title).description(Integer.toString(constraints[1])).descriptionEditable(true).descriptionEditInputType(InputType.TYPE_CLASS_NUMBER | InputType.TYPE_NUMBER_FLAG_SIGNED).build());
        // actions.add(new GuidedAction.Builder(activity).id(++i).title(R.string.min_rows_title).description(Integer.toString(constraints[0])).descriptionEditable(true).descriptionEditInputType(InputType.TYPE_CLASS_NUMBER | InputType.TYPE_NUMBER_FLAG_SIGNED).build());

        musicIndex = ((i - 1) / 2);

        // VIDEO
        state = categories.contains(AppCategory.VIDEO);
        statelabel = (state) ? getString(R.string.v7_preference_on) : getString(R.string.v7_preference_off);
        actions.add(new GuidedAction.Builder(activity).id(++i).title(R.string.videos_row_title).description(statelabel).build());
        // actions.add(new GuidedAction.Builder(activity).id(++i).checkSetId(i).checked(state).title(statelabel).description("").build());

        constraints = RowPreferences.getRowConstraints(AppCategory.VIDEO, activity);

        actions.add(new GuidedAction.Builder(activity).id(++i).title(R.string.max_rows_title).description(Integer.toString(constraints[1])).descriptionEditable(true).descriptionEditInputType(InputType.TYPE_CLASS_NUMBER | InputType.TYPE_NUMBER_FLAG_SIGNED).build());
        // actions.add(new GuidedAction.Builder(activity).id(++i).title(R.string.min_rows_title).description(Integer.toString(constraints[0])).descriptionEditable(true).descriptionEditInputType(InputType.TYPE_CLASS_NUMBER | InputType.TYPE_NUMBER_FLAG_SIGNED).build());

        videoIndex = ((i - 1) / 2);

        // GAME
        state = categories.contains(AppCategory.GAME);
        statelabel = (state) ? getString(R.string.v7_preference_on) : getString(R.string.v7_preference_off);
        actions.add(new GuidedAction.Builder(activity).id(++i).title(R.string.games_row_title).description(statelabel).build());
        // actions.add(new GuidedAction.Builder(activity).id(++i).checkSetId(i).checked(state).title(statelabel).description("").build());

        constraints = RowPreferences.getRowConstraints(AppCategory.GAME, activity);

        actions.add(new GuidedAction.Builder(activity).id(++i).title(R.string.max_rows_title).description(Integer.toString(constraints[1])).descriptionEditable(true).descriptionEditInputType(InputType.TYPE_CLASS_NUMBER | InputType.TYPE_NUMBER_FLAG_SIGNED).build());
        // actions.add(new GuidedAction.Builder(activity).id(++i).title(R.string.min_rows_title).description(Integer.toString(constraints[0])).descriptionEditable(true).descriptionEditInputType(InputType.TYPE_CLASS_NUMBER | InputType.TYPE_NUMBER_FLAG_SIGNED).build());

        gameIndex = ((i - 1) / 2);

        // ALL
        actions.add(new GuidedAction.Builder(activity).id(ACTION_ID_APPS).title(R.string.apps_row_title).build());
        constraints = RowPreferences.getAllAppsConstraints(activity);
        actions.add(new GuidedAction.Builder(activity).id(ACTION_ID_APPS_MAX).title(R.string.max_rows_title).description(Integer.toString(constraints[1])).descriptionEditable(true).descriptionEditInputType(InputType.TYPE_CLASS_NUMBER | InputType.TYPE_NUMBER_FLAG_SIGNED).build());
        actions.add(new GuidedAction.Builder(activity).id(ACTION_ID_APPS_MIN).title(R.string.min_rows_title).description(Integer.toString(constraints[0])).descriptionEditable(true).descriptionEditInputType(InputType.TYPE_CLASS_NUMBER | InputType.TYPE_NUMBER_FLAG_SIGNED).build());

        // INPUTS
        state = RowPreferences.areInputsEnabled(activity);
        statelabel = (state) ? getString(R.string.v7_preference_on) : getString(R.string.v7_preference_off);
        actions.add(new GuidedAction.Builder(activity).id(ACTION_ID_INPUTS).title(R.string.inputs_row_title).description(statelabel).build());
        // actions.add(new GuidedAction.Builder(activity).id(ACTION_ID_INPUTS).checkSetId(ACTION_ID_INPUTS).checked(state).title(statelabel).description("").build());

        setActions(actions); // APPLY

        // REFRESH HOME BC
        if (startup > 0) {
        	// Log.d("### CN", MainActivity.class.getName());
        	// Intent Broadcast = new Intent("com.amazon.tv.leanbacklauncher.MainActivity");
        	Intent Broadcast = new Intent(MainActivity.class.getName()); // ACTION
        	Broadcast.putExtra("RefreshHome", true);
        	activity.sendBroadcast(Broadcast);
        }
    }
}

File: app/src/main/java/com/amazon/tv/leanbacklauncher/settings/LegacyAppRowPreferenceFragment.kt
package com.amazon.tv.leanbacklauncher.settings

import android.app.Activity
import android.os.Build
import android.os.Bundle
import android.text.InputType
import android.widget.Toast
import androidx.core.content.res.ResourcesCompat
import androidx.leanback.app.GuidedStepSupportFragment
import androidx.leanback.widget.GuidanceStylist.Guidance
import androidx.leanback.widget.GuidedAction
import com.amazon.tv.firetv.leanbacklauncher.apps.AppCategory
import com.amazon.tv.firetv.leanbacklauncher.apps.RowPreferences.areFavoritesEnabled
import com.amazon.tv.firetv.leanbacklauncher.apps.RowPreferences.areInputsEnabled
import com.amazon.tv.firetv.leanbacklauncher.apps.RowPreferences.areRecommendationsEnabled
import com.amazon.tv.firetv.leanbacklauncher.apps.RowPreferences.getAppsColumns
import com.amazon.tv.firetv.leanbacklauncher.apps.RowPreferences.getEnabledCategories
import com.amazon.tv.firetv.leanbacklauncher.apps.RowPreferences.getFavoriteRowMax
import com.amazon.tv.firetv.leanbacklauncher.apps.RowPreferences.getRowMax
import com.amazon.tv.firetv.leanbacklauncher.apps.RowPreferences.setAppsColumns
import com.amazon.tv.firetv.leanbacklauncher.apps.RowPreferences.setFavoriteRowMax
import com.amazon.tv.firetv.leanbacklauncher.apps.RowPreferences.setFavoritesEnabled
import com.amazon.tv.firetv.leanbacklauncher.apps.RowPreferences.setGamesEnabled
import com.amazon.tv.firetv.leanbacklauncher.apps.RowPreferences.setInputsEnabled
import com.amazon.tv.firetv.leanbacklauncher.apps.RowPreferences.setMusicEnabled
import com.amazon.tv.firetv.leanbacklauncher.apps.RowPreferences.setRecommendationsEnabled
import com.amazon.tv.firetv.leanbacklauncher.apps.RowPreferences.setRowMax
import com.amazon.tv.firetv.leanbacklauncher.apps.RowPreferences.setVideosEnabled
import com.amazon.tv.firetv.leanbacklauncher.util.FireTVUtils.isAmazonNotificationsEnabled
import com.amazon.tv.leanbacklauncher.R
import com.amazon.tv.leanbacklauncher.util.Util.refreshHome
import java.util.*

/**
 * Created by rockon999 on 3/25/18.
 */
class LegacyAppRowPreferenceFragment : GuidedStepSupportFragment() {

    override fun onCreateGuidance(savedInstanceState: Bundle?): Guidance {
        return Guidance(
            getString(R.string.edit_row_title),  // title
            getString(R.string.select_app_customize_rows_title),  // description
            getString(R.string.settings_dialog_title),  // breadcrumb (home_screen_order_action_title)
            ResourcesCompat.getDrawable(resources, R.drawable.ic_settings_home, null)
        )
    }

    override fun onResume() {
        super.onResume()
        updateActions()
        startup++
    }

    override fun onPause() {
        super.onPause()
        startup = 0
    }

    private var musicIndex = 0
    private var gameIndex = 0
    private var videoIndex = 0
    private var favIndex = 0

    override fun onGuidedActionClicked(action: GuidedAction) {
        val id = action.id
        val catId = (id - 1) / 2
        val subId = ((id - 1) % 2).toInt()
        val enabled: Boolean
        val activity: Activity? = activity
        val categories = getEnabledCategories(activity!!)
        val value: Int

        if (catId == favIndex.toLong()) {
            when (subId) {
                0 -> {
                    enabled = areFavoritesEnabled(activity)
                    setFavoritesEnabled(activity, !enabled)
                }
                1 -> {
                    value = try {
                        action.description.toString().toInt()
                    } catch (nfe: NumberFormatException) {
                        1
                    }
                    setFavoriteRowMax(activity, value)
                }
            }
        } else if (catId == musicIndex.toLong()) {
            when (subId) {
                0 -> {
                    enabled = categories.contains(AppCategory.MUSIC)
                    setMusicEnabled(activity, !enabled)
                }
                1 -> {
                    value = try {
                        action.description.toString().toInt()
                    } catch (nfe: NumberFormatException) {
                        1
                    }
                    setRowMax(AppCategory.MUSIC, activity, value)
                }
            }
        } else if (catId == videoIndex.toLong()) {
            when (subId) {
                0 -> {
                    enabled = categories.contains(AppCategory.VIDEO)
                    setVideosEnabled(activity, !enabled)
                }
                1 -> {
                    value = try {
                        action.description.toString().toInt()
                    } catch (nfe: NumberFormatException) {
                        1
                    }
                    setRowMax(AppCategory.VIDEO, activity, value)
                }
            }
        } else if (catId == gameIndex.toLong()) {
            when (subId) {
                0 -> {
                    enabled = categories.contains(AppCategory.GAME)
                    setGamesEnabled(activity, !enabled)
                }
                1 -> {
                    value = try {
                        action.description.toString().toInt()
                    } catch (nfe: NumberFormatException) {
                        1
                    }
                    setRowMax(AppCategory.GAME, activity, value)
                }
            }
        } else if (id == ACTION_ID_APPS_MAX.toLong()) {
            value = try {
                action.description.toString().toInt()
            } catch (nfe: NumberFormatException) {
                1
            }
            setRowMax(AppCategory.OTHER, activity, value)
        } else if (id == ACTION_ID_APPS_COL.toLong()) {
            value = try {
                action.description.toString().toInt()
            } catch (nfe: NumberFormatException) {
                1
            }
            setAppsColumns(activity, value)
        } else if (id == ACTION_ID_RECOMMENDATIONS.toLong()) { // RECOMMENDATIONS
            enabled = areRecommendationsEnabled(activity)
            setRecommendationsEnabled(activity, !enabled)
            if (!enabled && isAmazonNotificationsEnabled(activity)) {
                Toast.makeText(
                    activity,
                    activity.getString(R.string.recs_warning_sale),
                    Toast.LENGTH_LONG
                ).show()
            }
        } else if (id == ACTION_ID_INPUTS.toLong()) { // INPUTS
            enabled = areInputsEnabled(activity)
            setInputsEnabled(activity, !enabled)
        }
        updateActions()
    }

    private fun updateActions() {

        val actions = ArrayList<GuidedAction>()
        val activity: Activity? = activity
        val categories = getEnabledCategories(activity!!)
        var desc: String
        var index = 0
        var state: Boolean

        // RECOMMENDATIONS
        if (Build.VERSION.SDK_INT < Build.VERSION_CODES.O) {
            state = areRecommendationsEnabled(activity)
            desc =
                if (state) getString(R.string.v7_preference_on) else getString(R.string.v7_preference_off)
            actions.add(
                GuidedAction.Builder(activity)
                    .id(ACTION_ID_RECOMMENDATIONS.toLong())
                    .title(R.string.recs_row_title)
                    .description(desc)
                    .build()
            )
        }
        // INPUTS
        state = areInputsEnabled(activity)
        desc =
            if (state) getString(R.string.v7_preference_on) else getString(R.string.v7_preference_off)
        actions.add(
            GuidedAction.Builder(activity)
                .id(ACTION_ID_INPUTS.toLong())
                .title(R.string.inputs_row_title)
                .description(desc)
                .build()
        )
        // FAV
        state = areFavoritesEnabled(activity)
        desc =
            if (state) getString(R.string.v7_preference_on) else getString(R.string.v7_preference_off)
        actions.add(
            GuidedAction.Builder(activity)
                .id((++index).toLong())
                .title(R.string.favorites_row_title)
                .description(desc)
                .build()
        )
        var maxRows = getFavoriteRowMax(activity)
        actions.add(
            GuidedAction.Builder(activity)
                .id((++index).toLong())
                .title(R.string.max_favorites_rows_title)
                .description(maxRows.toString())
                .descriptionEditable(true)
                .descriptionEditInputType(InputType.TYPE_CLASS_NUMBER or InputType.TYPE_NUMBER_FLAG_SIGNED)
                .build()
        )
        favIndex = (index - 1) / 2
        // VIDEO
        state = categories.contains(AppCategory.VIDEO)
        desc =
            if (state) getString(R.string.v7_preference_on) else getString(R.string.v7_preference_off)
        actions.add(
            GuidedAction.Builder(activity)
                .id((++index).toLong())
                .title(R.string.videos_row_title)
                .description(desc)
                .build()
        )
        maxRows = getRowMax(AppCategory.VIDEO, activity)
        actions.add(
            GuidedAction.Builder(activity)
                .id((++index).toLong())
                .title(R.string.max_videos_rows_title)
                .description(maxRows.toString())
                .descriptionEditable(true)
                .descriptionEditInputType(InputType.TYPE_CLASS_NUMBER or InputType.TYPE_NUMBER_FLAG_SIGNED)
                .build()
        )
        videoIndex = (index - 1) / 2
        // MUSIC
        state = categories.contains(AppCategory.MUSIC)
        desc =
            if (state) getString(R.string.v7_preference_on) else getString(R.string.v7_preference_off)
        actions.add(
            GuidedAction.Builder(activity)
                .id((++index).toLong())
                .title(R.string.music_row_title)
                .description(desc)
                .build()
        )
        maxRows = getRowMax(AppCategory.MUSIC, activity)
        actions.add(
            GuidedAction.Builder(activity)
                .id((++index).toLong())
                .title(R.string.max_music_rows_title)
                .description(maxRows.toString())
                .descriptionEditable(true)
                .descriptionEditInputType(InputType.TYPE_CLASS_NUMBER or InputType.TYPE_NUMBER_FLAG_SIGNED)
                .build()
        )
        musicIndex = (index - 1) / 2
        // GAME
        state = categories.contains(AppCategory.GAME)
        desc =
            if (state) getString(R.string.v7_preference_on) else getString(R.string.v7_preference_off)
        actions.add(
            GuidedAction.Builder(activity)
                .id((++index).toLong())
                .title(R.string.games_row_title)
                .description(desc)
                .build()
        )
        maxRows = getRowMax(AppCategory.GAME, activity)
        actions.add(
            GuidedAction.Builder(activity)
                .id((++index).toLong())
                .title(R.string.max_games_rows_title)
                .description(maxRows.toString())
                .descriptionEditable(true)
                .descriptionEditInputType(InputType.TYPE_CLASS_NUMBER or InputType.TYPE_NUMBER_FLAG_SIGNED)
                .build()
        )
        gameIndex = (index - 1) / 2
        // ALL APPS
        maxRows = getRowMax(AppCategory.OTHER, activity)
        val maxApps = getAppsColumns(activity)
        actions.add(
            GuidedAction.Builder(activity)
                .id(ACTION_ID_APPS_MAX.toLong())
                .title(R.string.max_apps_rows_title)
                .description(maxRows.toString())
                .descriptionEditable(true)
                .descriptionEditInputType(InputType.TYPE_CLASS_NUMBER or InputType.TYPE_NUMBER_FLAG_SIGNED)
                .build()
        )
        // Max Apps per row
        actions.add(
            GuidedAction.Builder(activity)
                .id(ACTION_ID_APPS_COL.toLong())
                .title(R.string.max_col_title)
                .description(maxApps.toString())
                .descriptionEditable(true)
                .descriptionEditInputType(InputType.TYPE_CLASS_NUMBER or InputType.TYPE_NUMBER_FLAG_SIGNED)
                .build()
        )
        setActions(actions) // APPLY

        // REFRESH HOME BC
        if (startup > 0) {
            refreshHome(requireActivity())
        }
    }

    companion object {
        //        private const val ACTION_ID_APPS = 50
        private const val ACTION_ID_APPS_COL = 51
        private const val ACTION_ID_APPS_MAX = 52
        private const val ACTION_ID_RECOMMENDATIONS = 100
        private const val ACTION_ID_INPUTS = 200
        private var startup = 0
    }
}